stock_mts_mto_rule/models/stock_rule.py

```diff
@@ -54,30 +54,51 @@ def get_mto_qty_to_order(self, product, product_qty, product_uom, values):
                 return product_qty - qty_available
         return product_qty
 
-    def _run_split_procurement(
-        self, product_id, product_qty, product_uom, location_id, name, origin, values
-    ):
+    def _run_split_procurement(self, procurements):
         precision = self.env["decimal.precision"].precision_get(
             "Product Unit of Measure"
         )
-        needed_qty = self.get_mto_qty_to_order(
-            product_id, product_qty, product_uom, values
-        )
-
-        if float_is_zero(needed_qty, precision_digits=precision):
-            getattr(self.mts_rule_id, "_run_%s" % self.mts_rule_id.action)(
-                product_id, product_qty, product_uom, location_id, name, origin, values
-            )
-        elif float_compare(needed_qty, product_qty, precision_digits=precision) == 0.0:
-            getattr(self.mto_rule_id, "_run_%s" % self.mto_rule_id.action)(
-                product_id, product_qty, product_uom, location_id, name, origin, values
-            )
-        else:
-            mts_qty = product_qty - needed_qty
-            getattr(self.mts_rule_id, "_run_%s" % self.mts_rule_id.action)(
-                product_id, mts_qty, product_uom, location_id, name, origin, values
-            )
-            getattr(self.mto_rule_id, "_run_%s" % self.mto_rule_id.action)(
-                product_id, needed_qty, product_uom, location_id, name, origin, values
+        for procurement, rule in procurements:
+            needed_qty = rule.get_mto_qty_to_order(
+                procurement.product_id,
+                procurement.product_qty,
+                procurement.product_uom,
+                procurement.values,
             )
+            if float_is_zero(needed_qty, precision_digits=precision):
+                getattr(self.env["stock.rule"], "_run_%s" % rule.mts_rule_id.action)(
+                    [(procurement, rule.mts_rule_id)]
+                )
+            elif (
+                float_compare(
+                    needed_qty, procurement.product_qty, precision_digits=precision
+                )
+                == 0.0
+            ):
+                getattr(self.env["stock.rule"], "_run_%s" % rule.mto_rule_id.action)(
+                    [(procurement, rule.mto_rule_id)]
+                )
+            else:
+                mts_qty = procurement.product_qty - needed_qty
+                mts_procurement = procurement._replace(product_qty=mts_qty)
+                getattr(self.env["stock.rule"], "_run_%s" % rule.mts_rule_id.action)(
+                    [(mts_procurement, rule.mts_rule_id)]
+                )
+
+                # Search all confirmed stock_moves of mts_procuremet and assign them
+                # to adjust the product's free qty
+                domain = self.env["procurement.group"]._get_moves_to_assign_domain()
```

Comment:
Is it possible to make this real multi implementation ? If procurements is real big recordset, you will do a search per record.

So, it's better to build domain before the loop. and assign moves at the end.

Reply:
Domain applied outside the cycle, however, it is necessary to assign the movements of procuremets_mts before applying the next procurement _mto